Add a timeout of 20 seconds to all executed commands.

dmitshur opened the pull request:
This should help if a remote server stalls and does not hang up in a long time.

Use a pretty conservative value of 10 seconds for timeout. `vcsstate` performs very lightweight operations, so 10 seconds should be more than enough time.

dmitshur committed:
Add a timeout of 10 seconds to all executed commands.

This should help if a remote server stalls and does not hang up in a
long time.

Use a pretty conservative value of 10 seconds for timeout. vcsstate
performs very lightweight operations, so 10 seconds should be more than
enough time.

StabbyCutyou commented:
If the timeout is hit, should you not return an error up the stack?

dmitshur commented:
I do. If a process is terminated via TERM signal, then `cmd.Wait()` will return non-clean exit status error, which gets returned from `outputTimeout`.

StabbyCutyou commented:
I suppose what I meant was, an explicit error about timeouts. That it
failed because of the timeout, so it's simpler to reason about.

Just a thought
On Apr 10, 2016 13:52, "Dmitri Shuralyov" notifications@github.com wrote:

I do. If a process is terminated via TERM signal, then cmd.Wait() will
return non-clean exit status error, which gets returned from outputTimeout.

—
You are receiving this because you commented.
Reply to this email directly or view it on GitHub
https://github.com/shurcooL/vcsstate/pull/2#issuecomment-208031526

dmitshur commented:
Right now `vcsstate` methods return one of two general types of results, success and failure. If a client cares about different types of failures, they can still inspect the returned `err` and see something different.

Right now `vcsstate` is [used by 5 clients](https://godoc.org/github.com/shurcooL/vcsstate?importers), they're all mine, and none of them care to distinguish between the reasons for failure. A timeout is not that special of a failure, so I don't see why it should be treated specially.

Maybe when I see a concrete need for it, I can consider doing something more, but until then, this seems fine.

I noticed that I forgot to update the two `RemoteVCS` implementations in `remotegit.go` and `remotehg.go`, I will fix that.

StabbyCutyou commented:
Sounds fine, just raising it as a question.

Thanks
On Apr 10, 2016 17:09, "Dmitri Shuralyov" notifications@github.com wrote:

> Right now vcsstate methods return one of two general types of results,
> success and failure. If a client cares about different types of failures,
> they can still inspect the returned err and see something different.
> 
> Right now vcsstate is used by 5 clients, they're all mine, and none of
> them care to distinguish between the reasons for failure. A timeout is not
> that special of a failure, so I don't see why it should be treated
> differently.
> 
> Maybe when I see a concrete need for it, I can consider doing something
> more, but until then, this seems fine.
> 
> I noticed that I forgot to update the two RemoteVCS implementations in
> remotegit.go and remotehg.go, I will fix that.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> https://github.com/shurcooL/vcsstate/pull/2#issuecomment-208071445

dmitshur commented:
Thanks, that's appreciated.

dmitshur committed:
Use os.Interrupt over syscall.SIGTERM.

Avoid using a non-portable and deprecated syscall package. Interrupt
signal is simpler and a better fit for this task.

dmitshur commented:
I'm also starting to feel 10 seconds might be a little too short. Establishing a TLS connection may sometimes take a few moments, and due to "turbulence" or whatnot, I can see it somehow taking 11 seconds in some rare cases.

Maybe I'll bump up the timeout to 20 seconds to be more confident there won't be false positives.

Going to sit on this PR for a while and think more about it.

dmitshur committed:
Increase timeout to 20 seconds to be more confident of no false positives.

Establishing TLS connections can sometimes be a little unpredictable
and lengthy... but 20 seconds should be enough to be confident it's not
going to work.

dmitshur commented:
I'll close this because I don't feel confident enough that doing this (in a configuration-free way) is the responsibility of this library, or that the constant values I picked are good, and that it's needed at all.

I can re-open when/if a concrete needs arises.
